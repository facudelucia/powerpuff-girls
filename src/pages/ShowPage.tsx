import React from 'react';
import ShowDetails from '../components/ShowDetails';
import styled from 'styled-components';

const PageContainer = styled.div`
  padding: 40px;
  display: flex;
  justify-content: center;
  align-items: center;
  min-height: 100vh;
  background: linear-gradient(135deg, #f5f7fa, #c3cfe2);
  
  @media (max-width: 768px) {
    padding: 20px;
  }
`;

const ShowCard = styled.div`
  background-color: white;
  padding: 30px;
  border-radius: 10px;
  box-shadow: 0 4px 20px rgba(0, 0, 0, 0.1);
  max-width: 1000px;
  width: 100%;
  transition: transform 0.3s ease, box-shadow 0.3s ease;
  
  &:hover {
    box-shadow: 0 6px 25px rgba(0, 0, 0, 0.15);
  }

  @media (max-width: 768px) {
    padding: 20px;
  }
`;

const ShowPage: React.FC = () => {
  return (
    <PageContainer>
      <ShowCard>
        <ShowDetails />
      </ShowCard>
    </PageContainer>
  );
};

export default ShowPage;
