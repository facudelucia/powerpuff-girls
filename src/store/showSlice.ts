import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import axios from "axios";
import { Show, Episode } from "../types/showTypes";

export interface ShowState {
  show: Show | null;
  episodes: Episode[];
  selectedEpisode: Episode | null;
  status: "idle" | "loading" | "failed";
}

const initialState: ShowState = {
  show: null,
  episodes: [],
  selectedEpisode: null,
  status: "idle",
};

export const fetchShowDetails = createAsyncThunk(
  "show/fetchShowDetails",
  async () => {
    const response = await axios.get("https://api.tvmaze.com/shows/6771");
    return response.data;
  }
);

export const fetchEpisodes = createAsyncThunk(
  "show/fetchEpisodes",
  async () => {
    const response = await axios.get(
      "https://api.tvmaze.com/shows/6771/episodes"
    );
    return response.data;
  }
);

const showSlice = createSlice({
  name: "show",
  initialState,
  reducers: {
    selectEpisode: (state, action) => {
      state.selectedEpisode =
        state.episodes.find(
          (episode: Episode) => episode.id === action.payload
        ) || null;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchShowDetails.pending, (state) => {
        state.status = "loading";
      })
      .addCase(fetchShowDetails.fulfilled, (state, action) => {
        state.status = "idle";
        state.show = action.payload;
      })
      .addCase(fetchEpisodes.fulfilled, (state, action) => {
        state.episodes = action.payload;
      });
  },
});

export const { selectEpisode } = showSlice.actions;
export default showSlice.reducer;
