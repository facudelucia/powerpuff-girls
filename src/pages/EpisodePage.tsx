import React, { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { useParams, useNavigate  } from 'react-router-dom';
import { selectEpisode, fetchEpisodes } from '../store/showSlice';
import EpisodeDetails from '../components/EpisodeDetails';
import { AppDispatch } from '../store';
import styled from 'styled-components';

const PageContainer = styled.div`
  padding: 40px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  min-height: 100vh;
  background: linear-gradient(135deg, #ece9e6, #ffffff);

  @media (max-width: 768px) {
    padding: 20px;
  }
`;

const EpisodeCard = styled.div`
  background-color: white;
  padding: 30px;
  border-radius: 10px;
  box-shadow: 0 4px 20px rgba(0, 0, 0, 0.1);
  max-width: 1000px;
  width: 100%;
  transition: transform 0.3s ease, box-shadow 0.3s ease;

  &:hover {
    box-shadow: 0 6px 25px rgba(0, 0, 0, 0.15);
  }

  @media (max-width: 768px) {
    padding: 20px;
  }
`;

const GoBackButton = styled.button`
  background-color: #3498db;
  color: white;
  padding: 10px 20px;
  font-size: 1rem;
  font-weight: bold;
  border: none;
  border-radius: 5px;
  cursor: pointer;
  margin-bottom: 20px;
  transition: background-color 0.3s ease, transform 0.2s ease;

  &:hover {
    background-color: #2980b9;
    transform: scale(1.05);
  }

  &:focus {
    outline: none;
  }
`;

const EpisodePage: React.FC = () => {
  const { id } = useParams<{ id: string }>();
  const dispatch: AppDispatch = useDispatch();
  const navigate = useNavigate(); 

  useEffect(() => {
      dispatch(selectEpisode(Number(id)));
  }, [dispatch, id]); 
  
  const handleGoBack = () => {
    navigate(-1);
  };

  return (
    <PageContainer>
      <EpisodeCard data-testid="episode-details">
        <GoBackButton onClick={handleGoBack}>{"< Go Back"}</GoBackButton>
        <EpisodeDetails />
      </EpisodeCard>
    </PageContainer>
  );
};

export default EpisodePage;
