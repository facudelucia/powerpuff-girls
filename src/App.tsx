import React from 'react';
import { BrowserRouter as Router, Route, Routes } from 'react-router-dom';
import ShowPage from './pages/ShowPage';
import EpisodePage from './pages/EpisodePage';

const App: React.FC = () => {
  return (
    <Router>
      <Routes>
        <Route path="/" element={<ShowPage />} />
        <Route path="/episode/:id" element={<EpisodePage />} />
      </Routes>
    </Router>
  );
};

export default App;
