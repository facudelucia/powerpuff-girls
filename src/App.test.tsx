import { render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import App from './App';
import ShowPage from './pages/ShowPage';
import EpisodePage from './pages/EpisodePage';
import { store } from './store';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer'

describe('App Component', () => {
  test('renders ShowPage component as the default route', async () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/']}>
          <Routes>
            <Route path="/" element={<ShowPage />} />
          </Routes>
        </MemoryRouter>
      </Provider>
    );

    await waitFor(() => {
      expect(screen.getByTestId('show-details')).toBeInTheDocument();
    });
  });

  test('renders EpisodePage component when navigating to /episode/:id', async () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/episode/657308']}>
          <Routes>
            <Route path="/episode/:id" element={<EpisodePage />} />
          </Routes>
        </MemoryRouter>
      </Provider>
    );

    await waitFor(() => {
      expect(screen.getByTestId('episode-details')).toBeInTheDocument();
    });
  });

  test('matches the snapshot', () => {
    const tree = renderer.create(
      <Provider store={store}>
          <App />
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
