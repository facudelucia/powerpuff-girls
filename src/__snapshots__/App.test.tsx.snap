// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component matches the snapshot 1`] = `
<div
  className="sc-knesRu ilcbuR"
>
  <div
    className="sc-dnaUSb cfHNqg"
  >
    <div
      className="sc-braxZu fZSJS"
      data-testid="show-details"
    >
      <h1
        className="sc-gJhJTp fcsndE"
      >
        The Powerpuff Girls
      </h1>
      <img
        alt="The Powerpuff Girls"
        className="sc-elDIKY RBvBg"
        src="https://static.tvmaze.com/uploads/images/original_untouched/60/151357.jpg"
      />
      <div
        className="sc-fQpRED hMCgaB"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<p>The city of Townsville may be a beautiful, bustling metropolis, but don't be fooled! There's evil afoot! And only three things can keep the bad guys at bay: Blossom, Bubbles and Buttercup, three super-powered little girls, known to their fans (and villains everywhere) as <b>The Powerpuff Girls</b>. Juggling school, bedtimes, and beating up giant monsters may be daunting, but together the Powerpuff Girls are up to the task. Battling a who's who of evil, they show what it really means to \\"fight like a girl.\\"</p>",
          }
        }
      />
      <h2>
        Episodes
      </h2>
      <ul
        className="sc-dsLQwm jVDesc"
      />
    </div>
  </div>
</div>
`;
